smqtk_iqr/iqr/iqr_session.py

```diff
@@ -300,25 +317,67 @@ def refine(self) -> None:
 
         """
         with self.lock:
-            # combine pos/neg adjudications + added external data descriptors
+            # Combine pos/neg adjudications + added external data descriptors
             pos = [desc.vector() for desc in (self.positive_descriptors |
                                               self.external_positive_descriptors)]
-
             neg = [desc.vector() for desc in (self.negative_descriptors |
                                               self.external_negative_descriptors)]
 
             if not pos:
                 raise RuntimeError("Did not find at least one positive "
                                    "adjudication.")
 
-            self._log.debug("Ranking working set with %d pos and %d neg total "
-                            "examples.", len(pos), len(neg))
+            # Get working set descriptors
             pool_uids, pool_de = zip(*self.working_set.items())
             pool = [de.vector() for de in pool_de]
+            pool_mat = np.asarray(pool)
+            pool_de_mat = np.asarray(pool_de)
+
+            # Auto-select negative examples if none are given
+            if not neg:
+                neg_autoselect = set()
+                self._log.info(f"Auto-selecting negative examples. "
+                               f"({self.autoneg_select_ratio} per positive)")
+
+                # For each positive example, find the farthest descriptor
+                # from it to use as a negative example
+                for p in pos:
+                    # Compute distance between current positive example and
+                    # each of the working set descriptors
+                    np_distances = self.distance_metric(p, pool_mat)
+
+                    # get array of the indices of the K maximally distant elements where
+                    # `K = autoneg_select_ratio` and `K >= 1.`
+                    part_size: int = self.autoneg_select_ratio
+                    max_indices: Sequence[int] = np.argpartition(np_distances, -part_size)[-part_size:]
+
+                    neg_autoselect.update(pool_de_mat[max_indices])
+
+                self._log.debug(f"Auto-selected negative descriptors (before difference update) "
+                                f"[{len(neg_autoselect)}]: {neg_autoselect}")
+
+                # Remove any positive examples from auto-selected results
+                neg_autoselect.difference_update((self.positive_descriptors |
+                                                  self.external_positive_descriptors))
+
+                self._log.debug(f"Auto-selected negative descriptors (after difference update) "
+                                f"[{len(neg_autoselect)}]: {neg_autoselect}")
+
+                if not neg_autoselect:
+                    raise RuntimeError("Negative auto-selection failed. "
+                                       "Did not select any negative examples.")
```

Comment:
We should add a unit test to exercise this case if it is not too much trouble.

Reply:
Unit test added to `tests/iqr/test_iqr_session.py`.